mecanum_drive_controller/src/odometry.cpp

```diff
@@ -0,0 +1,147 @@
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/*
+ * Author: Luca Marchionni
+ * Author: Bence Magyar
+ * Author: Enrique Fernández
+ * Author: Paul Mathieu
+ */
+
+#include <mecanum_drive_controller/odometry.h>
+
+#include <tf/transform_datatypes.h>
+
+#include <boost/bind.hpp>
+
+namespace mecanum_drive_controller
+{
+
+namespace bacc = boost::accumulators;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+Odometry::Odometry(size_t velocity_rolling_window_size)
+: timestamp_(0.0)
+, x_(0.0)
+, y_(0.0)
+, heading_(0.0)
+, linearX_(0.0)
+, linearY_(0.0)
+, angular_(0.0)
+, wheels_k_(0.0)
+, wheels_radius_(0.0)
+, velocity_rolling_window_size_(velocity_rolling_window_size)
+, linearX_acc_(RollingWindow::window_size = velocity_rolling_window_size)
+, linearY_acc_(RollingWindow::window_size = velocity_rolling_window_size)
+, angular_acc_(RollingWindow::window_size = velocity_rolling_window_size)
+, integrate_fun_(boost::bind(&Odometry::integrateExact, this, _1, _2, _3))
+{
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+void Odometry::init(const ros::Time& time)
+{
+  // Reset accumulators:
+  linearX_acc_ = RollingMeanAcc(RollingWindow::window_size = velocity_rolling_window_size_);
+  linearY_acc_ = RollingMeanAcc(RollingWindow::window_size = velocity_rolling_window_size_);
+  angular_acc_ = RollingMeanAcc(RollingWindow::window_size = velocity_rolling_window_size_);
+
+  // Reset timestamp:
+  timestamp_ = time;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+bool Odometry::update(double wheel0_vel, double wheel1_vel, double wheel2_vel, double wheel3_vel, const ros::Time &time)
+{
+  /// We cannot estimate the speed with very small time intervals:
+  const double dt = (time - timestamp_).toSec();
+  if (dt < 0.0001)
```

Comment:
Since now the wheel velocities are actually velocities, you don't have to divide by `dt`, so you can get rid of this.